MAINT: move comparison operator special-handling out of ufunc parsing.

mhvk opened the pull request:
The argument parsing for ufuncs contains special logic to enable returning `NotImplemented` for comparison functions.  The ufuncs, however, should not have to know whether they were called via a python comparison operator, so this PR moves the logic to the operators, checking after ufunc failed
whether a `NotImplemented` should be returned.

Follow-up of #11260, in particular https://github.com/numpy/numpy/pull/11260#discussion_r193649848

mhvk committed:
MAINT: move comparison operator special-handling out of ufunc parsing.

The argument parsing for ufuncs contains special logic to enable returning
NotImplemented for comparison functions.  The ufuncs, however, should
not have to know whether they were called via a python comparison operator,
so this PR moves the logic to the operators, checking after ufunc failed
whether a NotImplemented should be returned.

mhvk commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1464,8 +1450,102 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+         * is counted as inflexble.  (This repeats work done in the ufunc,
+         * but OK to waste some time in an unlikely path.)
+         */
+        if (PyArray_GetArrayParamsFromObject(
```
I tried using the seemingly more logical `PyArray_FromAny(other, NULL, 0, 0, 0, NULL)` but somehow this returns a failure when other is a python string.

eric-wieser replied:
You probably need to `PyErr_Fetch` above, before you start calling any more functions - most python functions are not willing to run if an exception is actively being thrown

mhvk replied:
Ah, yes, that indeed does it!

eric-wieser commented:
This presumably means `np.equal("one", 1)` no longer returns `NotImplemented?

mhvk commented:
@eric-wieser - yes, indeed, the ufuncs themselves will no longer return `NotImplemented` - which means this would definitely need a note to that extent. So, not for 1.15!

eric-wieser commented:
Can you add some tests for ufuncs no longer returning NotImplemented?

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1269,6 +1270,26 @@ DEPRECATE_silence_error(const char *msg) {
+ * silenced error.
+ */
+NPY_NO_EXPORT int
+DEPRECATE_FUTUREWARNING_silence_error(const char *msg) {
```
Maybe better to just add a `PyObject *warning_type` argument to the original function, and use `PyErr_WarnEx(warning_type, msg, 1)` instead.

mhvk replied:
Since I needed the fetch the error anyway, the need for the new function disappeared...

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1464,8 +1450,102 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+        int other_type_num, other_is_flexible;
+        /*
+         * Determine whether other has a flexible dtype; here, inconvertible
+         * is counted as inflexble.  (This repeats work done in the ufunc,
```
typo

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1464,8 +1450,102 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+            return Py_NotImplemented;
+        }
+    else if (other_is_flexible) {
+            /*
```
weird indent

mattip commented:
It would be nice even at this early stage to document the user-visible changes and change the tests appropriately

mhvk commented:
OK, now simpler with `PyArray_FromAny` - and some tests that were ready to go enabled.

@mattip - agreed about documentation, but there isn't even a file for the 1.16 release notes yet... But the main effect is the one @eric-wieser noted: now the `ufunc`s themselves will error rather than return `NotImplemented` (surely a good thing...)

charris commented:
Just to emphasize, do not merge before the 1.15 branch.

mhvk commented:
OK, tests now all pass - further review much appreciated!

mhvk commented:
Rebased, so now this is just a single commit.

mhvk committed:
DOC: add release note for change in output of comparison ufuncs.

mhvk commented:
With 1.15 branches, I rebased this and added a release note for 1.16. If everyone is still happy with this, it would be nice to merge it, since I have other changes to the ufunc parsing in the pipeline (#11333 in particular).

mhvk commented:
@mattip, @eric-wieser - any further comments on this?

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1464,8 +1430,114 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+            Py_DECREF(array_other);
+        }
+        else {
+            PyErr_Clear(); /* we restore the original error if needed */
```
Would be clearer if `other_is_flexible` and `ndim_other` were assigned in this path, rather than initialized.

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1464,8 +1430,114 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+                 * For arrays, we emit a future deprecation warning.
+                 */
+                if ((ndim_other != 0 || PyArray_NDIM(self) != 0) &&
+                    DEPRECATE_FUTUREWARNING(
```
Would be  clearer with the "should deprecate" and "did raise" condition split into separate ifs

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1366,26 +1367,6 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
-             * Comparisons should raise errors when element-wise comparison
-             * is not possible.
-             */
-            /* 2015-05-14, 1.10 */
```
Don't drop this date

mhvk commented:
OK, thanks, did all three.

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1464,8 +1430,121 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+                        /*
+                         * In future, we should create a correctly shaped
+                         * array of bool. For now, a placeholder error.
+                         */
```
I think this should

* Appear before the `DEPRECATE_FUTUREWARNING` line
* say `placeholder warning`, not placeholder error.

mhvk replied:
It is just that if the deprecation is removed, the chained exception should be replaced (I wanted to have code that would at least work, but didn't feel like creating the array now...). I clarified this.

eric-wieser replied:
I'd say that's obvious, in the same way that the `return NULL` should also be replaced

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1464,8 +1430,121 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+        return Py_NotImplemented;
+    }
+    if (result == NULL) {
+        /*
```
Can you extract the contents of this long `if` to `_failed_compare_workaround(self, other, op)`, or something similar?

mhvk replied:
Done.

mhvk committed:
MAINT: move failure code to its own function.

mhvk commented:
OK, pushed an extra commit moving the failure path to its own function.

mattip commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1269,6 +1270,108 @@ DEPRECATE_silence_error(const char *msg) {
+         * ordering, so we emit a warning.
+         */
+#if !defined(NPY_PY3K)
+        if (DEPRECATE(
```
Date and version comment needed

mattip commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1269,6 +1270,108 @@ DEPRECATE_silence_error(const char *msg) {
+             * If other did not have a flexible dtype, the error cannot
+             * have been caused by a lack of implementation in the ufunc.
+             */
+            if (DEPRECATE(
```
Date and version comment needed

mattip commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1269,6 +1270,108 @@ DEPRECATE_silence_error(const char *msg) {
+             * array of bool.
+             */
+            if (ndim_other != 0 || PyArray_NDIM(self) != 0) {
+                if (DEPRECATE_FUTUREWARNING(
```
Date and version comment needed

mhvk commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1269,6 +1270,108 @@ DEPRECATE_silence_error(const char *msg) {
+    }
+    else {
+        /* LE, LT, GT, or GE with non-flexible other; just pass on error */
+        PyErr_Restore(exc, val, tb);
```
Yes, this makes a lot more sense.

mhvk commented:
@mattip, @eric-wieser - all implemented.

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1269,6 +1270,109 @@ DEPRECATE_silence_error(const char *msg) {
+
+    /* LE, LT, GT, or GE with non-flexible other; just pass on error */
+fail:
+    PyErr_Restore(exc, val, tb);
```
This line should be the `ChainExceptions`, not `PyErr_Restore`. The former implies the latter, not vice versa

eric-wieser replied:
A comment here saying "Reraise the original exception, or chain the new one" might be nice too

eric-wieser commented on doc/release/1.16.0-notes.rst:
```diff
@@ -37,3 +37,12 @@ Improvements
+------------------------------------------------------------------
+
+Previously, comparison ufuncs would return `NotImplemented` if their arguments
+had structured dtypes, to help the comparison operators deal with those.  This
```
Can you point out that the comparison operators continue to return `NotImplemented` as before?

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1269,6 +1270,109 @@ DEPRECATE_silence_error(const char *msg) {
+             * For scalars, returning NotImplemented is correct.
+             * For arrays, we emit a future deprecation warning.
+             * If that is removed, the failure jump should be replaced
+             * by the creation of a correctly shaped array of bool.
```
Perhaps just "when this warning is removed, it should return a correctly-shaped array of bool.

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1269,6 +1270,109 @@ DEPRECATE_silence_error(const char *msg) {
+        return Py_NotImplemented;
+    }
+
+    /* LE, LT, GT, or GE with non-flexible other; just pass on error */
```
I'd probably add an `else { goto fail; }` here, just to have somwhere obvious to tie the comment to. The compiler should optimize it just the same anyway

mhvk commented:
@eric-wieser - thanks for the further comments; all done.

eric-wieser reviewed (commented):
Implementation looks good now.

Did you just cut up the original long comment, or did you throw it out and reword it? I haven't yet checked to see if all of the important bits survived.

mhvk commented:
Yes, I reworded the long comment, removing parts not relevant for comparison operators. I think it now (even) more clearly states where we want to go.

mhvk commented:
As this was approved, I'll go ahead  and merge so I can can rebase some of the other PRs (which will be easier to look at without the long distracting stuff in `get_ufunc_arguments`)

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1269,6 +1270,115 @@ DEPRECATE_silence_error(const char *msg) {
+    }
+    else {
+        /* LE, LT, GT, or GE with non-flexible other; just pass on error */
+        goto fail;
```
I'm not clear on why this path doesn't return NotImplemented in py3

mhvk replied:
The reason is that the comparison is actually implemented, i.e., the error is real and we act just like we would do for `__add__` (note that the behaviour was the same before this PR).
